### PR TITLE
Add toArray and fromArray methods to the serializer

### PR DESCRIPTION
**Pull request by tystr**
This PR implements the solution suggested by @schmittjoh to add additional methods for dealing with array input/output instead of adding a new format for arrays.

https://github.com/schmittjoh/serializer/pull/20#issuecomment-94781669

TODO
- [x] Add some test cases

### CODE REVIEW

File: tests/JMS/Serializer/Tests/Serializer/ArrayTest.php [hunk 1]
```diff
@@ -0,0 +1,1307 @@
+<?php
```
Comment (schmittjoh) [inline tests/JMS/Serializer/Tests/Serializer/ArrayTest.php]: Let's not replicate the entire test suite. One simple test per method should be ok.

File: src/JMS/Serializer/Serializer.php [hunk 2]
```diff
@@ -121,6 +121,49 @@ public function deserialize($data, $type, $format, DeserializationContext $conte
...
+            'json',
+            $visitor = $this->serializationVisitors->get('json')->get(),
+            $this->navigator,
+            $this->factory
+        );
+
+        $visitor->setNavigator($this->navigator);
+        $this->navigator->accept($visitor->prepare($data), null, $context);
+
+        return $visitor->getRoot();
```
Comment (schmittjoh) [inline src/JMS/Serializer/Serializer.php]: We need some special handling for the ``ArrayObject`` here that the JSON visitor returns sometimes. Could you add a test for that, too?

**Comment by tystr** [comment]
I've just returned from vacation. I'll wrap this up this week.

**Comment by davidwindell** [comment]
:+1:

**Comment by tystr** [comment]
@schmittjoh I've simplified the tests for the array methods down to a couple of very rudimentary tests and added handling for `ArrayObject`.

File: src/JMS/Serializer/Serializer.php [hunk 3]
```diff
@@ -121,6 +121,49 @@ public function deserialize($data, $type, $format, DeserializationContext $conte
...
+            'json',
+            $visitor = $this->serializationVisitors->get('json')->get(),
+            $this->navigator,
+            $this->factory
+        );
+
+        $visitor->setNavigator($this->navigator);
+        $this->navigator->accept($visitor->prepare($data), null, $context);
+
+        return (array) $visitor->getRoot();
```
Comment (schmittjoh) [inline src/JMS/Serializer/Serializer.php]: We should do this for the entire array (including children) as these might have an array object, too.

File: tests/JMS/Serializer/Tests/Serializer/ArrayTest.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $this->serializer = new Serializer(
+            new MetadataFactory(new AnnotationDriver(new AnnotationReader())),
+            new HandlerRegistry(),
+            new UnserializeObjectConstructor(),
+            new Map(array('json' => new JsonSerializationVisitor($namingStrategy))),
+            new Map(array('json' => new JsonDeserializationVisitor($namingStrategy)))
+        );
+    }
+
+    public function testToArray()
```
Comment (schmittjoh) [inline tests/JMS/Serializer/Tests/Serializer/ArrayTest.php]: Could you add a test how we behave if for example a string or another scalar is passed? We should probably throw an exception if the result is not an array.

**Comment by rdohms** [comment]
:+1:

**Comment by rdohms** [comment]
Where are we with this? Any ideas when a release with it will come out?

**Comment by schmittjoh** [comment]
I've made a few tweaks, and merged this. We'll need to give this a bit of time in master before thinking about a release.

**Comment by davidwindell** [comment]
Great, thanks @tystr and @schmittjoh for getting this done!

**Comment by davidwindell** [comment]
Just a quick update, this has been used by our app in production since 1 May and all good so far!

**Comment by rdohms** [comment]
can we get this in a release?

**Comment by rdohms** [comment]
Aha, you did, v1.0.0 now supports this. :+1:

**Pull request merged**